plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -234,6 +256,129 @@ public void onFailure(Exception e) {
+            List<String> allDependentPipelineIds = findDependentPipelines(
+                ingestPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.ingest.PipelineConfiguration::getConfigAsMap,
+                org.opensearch.ingest.PipelineConfiguration::getId
+            );
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            allDependentPipelineIds.size()
+                                + " ingest pipelines are still using this model, please delete or update the pipelines first: "
+                                + Arrays.toString(allDependentPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkSearchPipelineBeforeDeleteModel(String modelId, ActionListener<Boolean> actionListener) {
+        GetSearchPipelineRequest getSearchPipelineRequest = new GetSearchPipelineRequest();
+        client.execute(GetSearchPipelineAction.INSTANCE, getSearchPipelineRequest, ActionListener.wrap(searchPipelineResponse -> {
+            List<String> allDependentPipelineIds = findDependentPipelines(
+                searchPipelineResponse.pipelines(),
+                modelId,
+                org.opensearch.search.pipeline.PipelineConfiguration::getConfigAsMap,
+                org.opensearch.search.pipeline.PipelineConfiguration::getId
+            );
+            if (allDependentPipelineIds.isEmpty()) {
+                actionListener.onResponse(true);
+            } else {
+                actionListener
+                    .onFailure(
+                        new OpenSearchStatusException(
+                            allDependentPipelineIds.size()
+                                + " search pipelines are still using this model, please delete or update the pipelines first: "
+                                + Arrays.toString(allDependentPipelineIds.toArray(new String[0])),
+                            RestStatus.CONFLICT
+                        )
+                    );
+            }
+        }, e -> {
+            log.error("Failed to delete ML Model: " + modelId, e);
+            actionListener.onFailure(e);
+
+        }));
+
+    }
+
+    private void checkDownstreamTaskBeforeDeleteModel(String modelId, Boolean isHidden, ActionListener<DeleteResponse> actionListener) {
+        CountDownLatch countDownLatch = new CountDownLatch(3);
+        AtomicBoolean noneBlocked = new AtomicBoolean(true);
+        List<String> errorMessages = new ArrayList<>();
```

Comment:
Let's change this ArrayList to a thread safe implementation as three requests could complete at same time and writing error messages concurrently, which might cause concurrent issue.